File: ClientTCP.java
/**
 * ClientTCP.java — Java TCP Chat Client
 * Computer Networks Lab
 *
 * Connects to the chat server over TCP (port 5556).
 * Two threads run concurrently:
 *   - receiver thread: prints messages pushed by the server
 *   - main thread:     reads stdin and sends messages to the server
 *
 * Compile:  javac ClientTCP.java
 * Run:      java ClientTCP
 */

import java.io.*;
import java.net.*;

public class ClientTCP {

    static final String HOST = "localhost";
    static final int    PORT = 5556;

    public static void main(String[] args) throws IOException {

        // Get display name from the user
        BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
        System.out.print("Enter your name: ");
        String username = stdin.readLine().trim();
        if (username.isEmpty()) username = "anonymous";

        System.out.println("[client] connecting to " + HOST + ":" + PORT + " …");

        // Perform the TCP three-way handshake
        Socket socket = new Socket(HOST, PORT);

        PrintWriter writer =
                new PrintWriter(new BufferedWriter(
                        new OutputStreamWriter(socket.getOutputStream())), true);
        BufferedReader serverReader =
                new BufferedReader(new InputStreamReader(socket.getInputStream()));

        System.out.println("[client] connected!  You are '" + username + "'");
        System.out.println("[client] type a message and press Enter.  'quit' to exit.\n");

        // First message registers our display name on the server
        writer.println(username);

        // ── receiver thread: print anything the server sends ──────────────────
        Thread receiverThread = new Thread(() -> {
            try {
                String line;
                while ((line = serverReader.readLine()) != null) {
                    // \r moves cursor to start of line so incoming messages
                    // don't collide visually with what the user is typing
                    System.out.print("\r" + line + "\n> ");
                }
            } catch (IOException e) {
                // Server closed the connection
            }
            System.out.println("\n[client] server closed the connection.");
        });
        receiverThread.setDaemon(true);   // exits automatically when main thread ends
        receiverThread.start();

        // ── main thread: read stdin and send to server ────────────────────────
        while (true) {
            System.out.print("> ");
            String line = stdin.readLine();
            if (line == null) break;              // EOF (Ctrl-D)
            line = line.trim();
            if (line.equalsIgnoreCase("quit") ||
                line.equalsIgnoreCase("exit") ||
                line.equalsIgnoreCase("q")) break;
            if (!line.isEmpty()) {
                writer.println(line);             // sends line + '\n' to server
            }
        }

        socket.close();
        System.out.println("\n[client] disconnected.");
    }
}

File: client_tcp.py
"""
client_tcp.py — TCP Chat Client
Computer Networks Lab

Connects to the chat server over TCP (port 5555).
Sends messages typed by the user and prints messages from other clients.

Usage:
    python client_tcp.py
"""

import asyncio
import sys


# ── entry point ───────────────────────────────────────────────────────────────

async def main():
    HOST = "localhost"
    PORT = 5555

    username = input("Enter your name: ").strip() or "anonymous"

    print(f"[client] connecting to {HOST}:{PORT} …")

    # asyncio.open_connection() performs the TCP three-way handshake and
    # returns a (StreamReader, StreamWriter) pair.
    reader, writer = await asyncio.open_connection(HOST, PORT)

    print(f"[client] connected!  You are '{username}'")
    print("[client] type a message and press Enter.  'quit' to exit.\n")

    # First message to the server sets our display name
    writer.write((username + "\n").encode())

    # ── receiver task: print messages pushed by the server ────────────────────
    async def receive_loop():
        while True:
            try:
                data = await reader.readline()
            except (asyncio.IncompleteReadError, ConnectionResetError):
                break
            if not data:
                break
            text = data.decode(errors="replace").strip()
            if text:
                print(f"\r{text}\n> ", end="", flush=True)
        print("\n[client] server closed the connection.")

    # ── sender task: read stdin and send lines to the server ──────────────────
    async def send_loop():
        loop = asyncio.get_event_loop()
        while True:
            print("> ", end="", flush=True)
            try:
                line = await loop.run_in_executor(None, sys.stdin.readline)
            except (EOFError, KeyboardInterrupt):
                break
            line = line.rstrip("\n").strip()
            if not line:
                continue
            if line.lower() in ("quit", "exit", "q"):
                break
            writer.write((line + "\n").encode())

    # Run both tasks concurrently; stop when either finishes
    done, pending = await asyncio.wait(
        [
            asyncio.create_task(receive_loop()),
            asyncio.create_task(send_loop()),
        ],
        return_when=asyncio.FIRST_COMPLETED,
    )
    for task in pending:
        task.cancel()

    try:
        writer.close()
        await writer.wait_closed()
    except Exception:
        pass

    print("\n[client] disconnected.")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n[client] disconnected.")
